include/boost/json/impl/monotonic_resource.ipp

```diff
@@ -18,38 +19,55 @@ namespace json {
 struct monotonic_resource::block
 {
     std::size_t const size;
+    block* next = nullptr;
     std::uintptr_t top;
-    block* next;
 
     block(
         std::size_t size_,
         block* next_)
         : size(size_)
-        , top(reinterpret_cast<
-            std::uintptr_t>(this+1))
         , next(next_)
+        , top(base_ptr())
     {
     }
 
-    void*
-    alloc(
-        std::size_t n,
-        std::size_t align) noexcept
+    std::uintptr_t
+    base_ptr() const
     {
-        // must be power of 2
-        BOOST_ASSERT(
-            (align & (align - 1)) == 0);
-        auto i = top;
-        if((i & (align - 1)) != 0)
-            i = (i | (align - 1)) + 1;
-        if(i + n > (reinterpret_cast<
-            std::uintptr_t>(this+1) + size))
-            return nullptr;
-        top = i + n;
-        return reinterpret_cast<void*>(i);
+        return reinterpret_cast<
+            std::uintptr_t>(this + 1);
     }
 };
 
+std::uintptr_t
+monotonic_resource::
+align_top(
+    std::uintptr_t ptr,
+    std::size_t align)
+{
+    // must be power of 2
+    BOOST_ASSERT(
+        (align & (align - 1)) == 0);
+    auto i = ptr;
+    return i & (align - 1) ?
+        (i | (align - 1)) + 1 : i;
+}
+
+template<typename Block>
+void*
+monotonic_resource::
+alloc_in_block(
```

Comment:
The "name" of a function includes the types of its parameters, so `alloc( Block& b, std::size_t n, std::size_t align )` says the same thing as `alloc_in_block( ...`